storage/zynq7000-flash/flashdrv.c

```diff
@@ -507,17 +483,58 @@ static int flashdrv_mtdErase(struct _storage_t *strg, off_t offs, size_t size)
 
 		len += secSz;
 	}
+
+	beg = offs - strg->start;
+	end = beg + size;
+
+	/* Invalidate block device cache for coherence */
+	res = cache_invalidate(fdrv_common.regs[strg->dev->ctx->id].cache, beg, end);
+	if (res < 0) {
+		mutexUnlock(fdrv_common.regs[strg->dev->ctx->id].lock);
+		return res;
+	}
+
 	mutexUnlock(fdrv_common.regs[strg->dev->ctx->id].lock);
 
 	return EOK;
 }
 
 
+/* Flushes and invalidates block device cache for coherence during MTD operations */
+static int _flashdrv_mtdFlushInv(struct _storage_t *strg, off_t beg, off_t end)
+{
+	int res;
+
+	res = cache_flush(fdrv_common.regs[strg->dev->ctx->id].cache, beg, end);
+	if (res < 0) {
+		return res;
+	}
+
+	res = cache_invalidate(fdrv_common.regs[strg->dev->ctx->id].cache, beg, end);
+	if (res < 0) {
+		return res;
+	}
+
+	return EOK;
+}
+
+
 static int flashdrv_mtdRead(struct _storage_t *strg, off_t offs, void *data, size_t len, size_t *retlen)
 {
 	int res;
+	off_t beg, end;
+
+	beg = offs - strg->start;
+	end = beg + len;
 
 	mutexLock(fdrv_common.regs[strg->dev->ctx->id].lock);
+
+	res = _flashdrv_mtdFlushInv(strg, beg, end);
```

Comment:
Hmm, I think cache invalidation should not be needed here, only flush is needed imho